File: code/runtime/hvm/ios/handle.swift
// handle.swift - Handle table for bridging Swift objects into HVM.
//
// Maps integer IDs to AnyObject references. ARC manages lifetime.
// When HVM fires ERA on a handle, call release to remove it from
// the table. If no other Swift code holds a reference, ARC
// deallocates the object.

final class HandleTable {
    private var nextId: UInt32 = 1
    private var objects: [UInt32: AnyObject] = [:]

    func register(_ obj: AnyObject) -> UInt32 {
        let id = nextId
        nextId += 1
        objects[id] = obj
        return id
    }

    func get(_ id: UInt32) -> AnyObject? {
        return objects[id]
    }

    func release(_ id: UInt32) {
        objects.removeValue(forKey: id)
    }

    func has(_ id: UInt32) -> Bool {
        return objects[id] != nil
    }

    func clear() {
        objects.removeAll()
        nextId = 1
    }

    var count: Int {
        return objects.count
    }
}

File: code/runtime/hvm/server/handle.rs
// handle.rs - Handle table for bridging Rust objects into HVM.
//
// Thread-safe via Mutex. Uses Arc for shared ownership.

use std::any::Any;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

pub struct HandleTable {
  next_id: u32,
  objects: HashMap<u32, Arc<dyn Any + Send + Sync>>,
}

impl HandleTable {
  pub fn new() -> Mutex<Self> {
    Mutex::new(Self {
      next_id: 1,
      objects: HashMap::new(),
    })
  }

  pub fn register(&mut self, obj: Arc<dyn Any + Send + Sync>) -> u32 {
    let id = self.next_id;
    self.next_id += 1;
    self.objects.insert(id, obj);
    id
  }

  pub fn get(&self, id: u32) -> Option<&Arc<dyn Any + Send + Sync>> {
    self.objects.get(&id)
  }

  pub fn release(&mut self, id: u32) {
    self.objects.remove(&id);
  }

  pub fn has(&self, id: u32) -> bool {
    self.objects.contains_key(&id)
  }

  pub fn clear(&mut self) {
    self.objects.clear();
    self.next_id = 1;
  }

  pub fn len(&self) -> usize {
    self.objects.len()
  }

  pub fn is_empty(&self) -> bool {
    self.objects.is_empty()
  }
}
